multiplex.go

```diff
@@ -430,33 +450,70 @@ func (mp *Multiplex) handleIncoming() {
 				// We're not accepting data on this stream, for
 				// some reason. It's likely that we reset it, or
 				// simply canceled reads (e.g., called Close).
-				mp.putBufferInbound(b)
+				if err := mp.skipNextMsg(mlen); err != nil {
+					mp.shutdownErr = err
+					return
+				}
 				continue
 			}
 
-			recvTimeout.Reset(ReceiveTimeout)
-			select {
-			case msch.dataIn <- b:
-			case <-msch.readCancel:
-				// the user has canceled reading. walk away.
-				mp.putBufferInbound(b)
-			case <-recvTimeout.C:
-				mp.putBufferInbound(b)
-				log.Warnf("timed out receiving message into stream queue.")
-				// Do not do this asynchronously. Otherwise, we
-				// could drop a message, then receive a message,
-				// then reset.
-				msch.Reset()
-				continue
-			case <-mp.shutdown:
-				mp.putBufferInbound(b)
-				return
-			}
-			if !recvTimeout.Stop() {
-				<-recvTimeout.C
+		read:
+			for rd := 0; rd < mlen; {
+				nextChunk := mlen - rd
+				if nextChunk > BufferSize {
+					nextChunk = BufferSize
+				}
+
+				b, err := mp.readNextMsg(nextChunk)
+				if err != nil {
+					mp.shutdownErr = err
+					return
+				}
+
+				rd += nextChunk
+
+				if !recvTimeout.Stop() {
+					select {
+					case <-recvTimeout.C:
```

Comment:
This doesn't do what you think it does.

Reply:
Basically, the timeout can fire immediately after this check and then we'll timeout on the read immediately.

Reply:
no, it has been stopped.

Reply:
1. If we can guarantee that the timer was started before we called stop, we should read and block.
2. If we can't guarantee that, we should find a way to guarantee that.

Reply:
1. We cant.
2. We could use a boolean variable indicating whether it fired.

side note: i really hate timers, they are a .45 footgun.

Reply:
fixed, using a variable to track whether the channel has been drained.